File: games/breakpoint-unity/Assets/BREAKPOINT/Runtime/Simulation/BallBody.cs
using System;

namespace Breakpoint.Simulation
{
    /// <summary>
    /// One ball's complete dynamic state.
    ///
    /// <see cref="Spin"/> is the full angular velocity vector in world axes:
    /// X and Y are the rolling axes (what topspin and backspin live on) and Z is
    /// English. Everything presentation needs to orient a ball comes from
    /// integrating this, so spin is never faked visually.
    ///
    /// A class rather than a struct: the world holds and mutates these in place
    /// through a whole shot, and copying a struct around would make it far too
    /// easy to update a temporary instead of the ball.
    /// </summary>
    public sealed class BallBody
    {
        /// <summary>Index in the world's ball list. Stable for the life of a shot.</summary>
        public readonly int Id;

        /// <summary>0 = cue ball, 1..15 = object balls.</summary>
        public readonly int Number;

        public Vec2 Position;
        public Vec2 Velocity;
        public Vec3 Spin;

        /// <summary>Accumulated orientation as a quaternion (x, y, z, w). Presentation only.</summary>
        public Quat Orientation;

        /// <summary>True once the ball has dropped; pocketed balls leave the simulation.</summary>
        public bool Pocketed;

        /// <summary>Which pocket it fell in, for the shot record.</summary>
        public string PocketId;

        /// <summary>True when both velocity and spin are exactly zero.</summary>
        public bool Resting;

        public BallBody(int id, int number, Vec2 position)
        {
            Id = id;
            Number = number;
            Position = position;
            Velocity = Vec2.Zero;
            Spin = Vec3.Zero;
            Orientation = Quat.Identity;
            Pocketed = false;
            PocketId = null;
            Resting = true;
        }

        public BallBody Clone()
        {
            return new BallBody(Id, Number, Position)
            {
                Velocity = Velocity,
                Spin = Spin,
                Orientation = Orientation,
                Pocketed = Pocketed,
                PocketId = PocketId,
                Resting = Resting,
            };
        }

        /// <summary>Copy another ball's dynamic state onto this one, in place.</summary>
        public void CopyStateFrom(BallBody other)
        {
            Position = other.Position;
            Velocity = other.Velocity;
            Spin = other.Spin;
            Orientation = other.Orientation;
            Pocketed = other.Pocketed;
            PocketId = other.PocketId;
            Resting = other.Resting;
        }

        /// <summary>
        /// Velocity of the contact patch where the ball touches the cloth, at
        /// r = (0, 0, −R) from the centre: u = v + ω × r.
        ///
        /// Expanding the cross product for that r gives
        /// u = (vx − R·ωy, vy + R·ωx). When u is zero the ball rolls without
        /// slipping; while it is non-zero the ball is sliding.
        /// </summary>
        public Vec2 ContactVelocity()
        {
            return new Vec2(
                Velocity.X - PhysicsConstants.BallRadius * Spin.Y,
                Velocity.Y + PhysicsConstants.BallRadius * Spin.X);
        }

        /// <summary>Is this ball below both rest thresholds?</summary>
        public bool BelowRestThreshold()
        {
            double v = Math.Sqrt(Velocity.X * Velocity.X + Velocity.Y * Velocity.Y);
            double w = Math.Sqrt(Spin.X * Spin.X + Spin.Y * Spin.Y + Spin.Z * Spin.Z);
            return v < PhysicsConstants.RestSpeed && w < PhysicsConstants.RestSpin;
        }

        /// <summary>Snap a near-stopped ball to exact rest so the shot terminates.</summary>
        public void ForceRest()
        {
            Velocity = Vec2.Zero;
            Spin = Vec3.Zero;
            Resting = true;
        }

        public bool IsFinite =>
            Position.IsFinite && Velocity.IsFinite && Spin.IsFinite;

        /// <summary>
        /// Roll the presentation orientation forward by ω·dt.
        ///
        /// This is bookkeeping on top of the real physics, not a substitute for
        /// it: the quaternion is derived from the same <see cref="Spin"/> that
        /// drives the trajectory, so the visible rotation and the physical
        /// rotation cannot disagree.
        /// </summary>
        public void IntegrateOrientation(double dt)
        {
            double wx = Spin.X;
            double wy = Spin.Y;
            double wz = Spin.Z;
            double magnitude = Math.Sqrt(wx * wx + wy * wy + wz * wz);
            if (magnitude < 1e-9) return;

            double half = magnitude * dt / 2.0;
            double s = Math.Sin(half) / magnitude;
            var delta = new Quat(wx * s, wy * s, wz * s, Math.Cos(half));
            Orientation = Quat.Multiply(delta, Orientation).Normalized();
        }
    }

    /// <summary>
    /// A quaternion, carried purely so presentation can show a ball rotating.
    /// The simulation never reads it back.
    /// </summary>
    public readonly struct Quat
    {
        public readonly double X;
        public readonly double Y;
        public readonly double Z;
        public readonly double W;

        public Quat(double x, double y, double z, double w)
        {
            X = x;
            Y = y;
            Z = z;
            W = w;
        }

        public static readonly Quat Identity = new Quat(0.0, 0.0, 0.0, 1.0);

        public bool IsFinite =>
            Numeric.IsFinite(X) && Numeric.IsFinite(Y) && Numeric.IsFinite(Z) && Numeric.IsFinite(W);

        public static Quat Multiply(Quat d, Quat q)
        {
            return new Quat(
                d.W * q.X + d.X * q.W + d.Y * q.Z - d.Z * q.Y,
                d.W * q.Y - d.X * q.Z + d.Y * q.W + d.Z * q.X,
                d.W * q.Z + d.X * q.Y - d.Y * q.X + d.Z * q.W,
                d.W * q.W - d.X * q.X - d.Y * q.Y - d.Z * q.Z);
        }

        public Quat Normalized()
        {
            double inv = 1.0 / Math.Sqrt(X * X + Y * Y + Z * Z + W * W);
            return new Quat(X * inv, Y * inv, Z * inv, W * inv);
        }
    }
}

File: games/breakpoint/src/physics/BallBody.ts
import { BALL_RADIUS, REST_SPEED, REST_SPIN } from './PhysicsConstants';
import { clone2, clone3, type Vec2, type Vec3 } from './Vec';

/**
 * One ball's complete dynamic state.
 *
 * `spin` is the full angular velocity vector in world axes: spin.x / spin.y are
 * the rolling axes (they are what topspin and backspin live on) and spin.z is
 * English. Everything the renderer needs to orient a ball comes from
 * integrating this, so spin is never faked visually.
 */
export interface BallBody {
  readonly id: number;
  /** 0 = cue ball, 1..15 = object balls. */
  readonly number: number;
  position: Vec2;
  velocity: Vec2;
  spin: Vec3;
  /** Accumulated orientation, as a quaternion (x, y, z, w). Render-only. */
  orientation: [number, number, number, number];
  /** True once the ball has dropped; pocketed balls leave the simulation. */
  pocketed: boolean;
  /** Which pocket it fell in, for the shot record. */
  pocketId: string | null;
  /** True when both velocity and spin are exactly zero. */
  resting: boolean;
}

export function createBall(id: number, number: number, position: Vec2): BallBody {
  return {
    id,
    number,
    position: clone2(position),
    velocity: { x: 0, y: 0 },
    spin: { x: 0, y: 0, z: 0 },
    orientation: [0, 0, 0, 1],
    pocketed: false,
    pocketId: null,
    resting: true,
  };
}

export function cloneBall(b: BallBody): BallBody {
  return {
    id: b.id,
    number: b.number,
    position: clone2(b.position),
    velocity: clone2(b.velocity),
    spin: clone3(b.spin),
    orientation: [...b.orientation] as [number, number, number, number],
    pocketed: b.pocketed,
    pocketId: b.pocketId,
    resting: b.resting,
  };
}

/** Linear kinetic energy, 1/2 m v². Mass is factored out by the caller. */
export function linearSpeedSq(b: BallBody): number {
  return b.velocity.x * b.velocity.x + b.velocity.y * b.velocity.y;
}

export function spinMagSq(b: BallBody): number {
  return b.spin.x * b.spin.x + b.spin.y * b.spin.y + b.spin.z * b.spin.z;
}

/**
 * Velocity of the contact patch where the ball touches the cloth, at
 * r = (0, 0, -R) from the centre:  u = v + ω × r.
 *
 * Expanding the cross product for r = (0,0,-R) gives u = (vx - R ωy, vy + R ωx).
 * When u is zero the ball rolls without slipping; while it is non-zero the ball
 * is sliding and sliding friction applies.
 */
export function contactVelocity(b: BallBody): Vec2 {
  return {
    x: b.velocity.x - BALL_RADIUS * b.spin.y,
    y: b.velocity.y + BALL_RADIUS * b.spin.x,
  };
}

/** Is this ball below both rest thresholds? */
export function belowRestThreshold(b: BallBody): boolean {
  const v = Math.hypot(b.velocity.x, b.velocity.y);
  const w = Math.hypot(b.spin.x, b.spin.y, b.spin.z);
  return v < REST_SPEED && w < REST_SPIN;
}

/** Snap a near-stopped ball to exact rest so the shot actually terminates. */
export function forceRest(b: BallBody): void {
  b.velocity.x = 0;
  b.velocity.y = 0;
  b.spin.x = 0;
  b.spin.y = 0;
  b.spin.z = 0;
  b.resting = true;
}

/**
 * Roll the render orientation forward by ω·dt.
 *
 * This is bookkeeping on top of the real physics, not a substitute for it: the
 * quaternion is derived from the same `spin` that drives the trajectory, so the
 * visible rotation and the physical rotation cannot disagree.
 */
export function integrateOrientation(b: BallBody, dt: number): void {
  const { x: wx, y: wy, z: wz } = b.spin;
  const mag = Math.hypot(wx, wy, wz);
  if (mag < 1e-9) return;
  const half = (mag * dt) / 2;
  const s = Math.sin(half) / mag;
  const dq: [number, number, number, number] = [wx * s, wy * s, wz * s, Math.cos(half)];
  const [qx, qy, qz, qw] = b.orientation;
  const [dx, dy, dz, dw] = dq;
  const nx = dw * qx + dx * qw + dy * qz - dz * qy;
  const ny = dw * qy - dx * qz + dy * qw + dz * qx;
  const nz = dw * qz + dx * qy - dy * qx + dz * qw;
  const nw = dw * qw - dx * qx - dy * qy - dz * qz;
  const inv = 1 / Math.hypot(nx, ny, nz, nw);
  b.orientation = [nx * inv, ny * inv, nz * inv, nw * inv];
}
